docs/design/scheduler_plugin.md

```diff
@@ -0,0 +1,112 @@
+significantly improving the compatibility of YuniKorn with upstream Kubernetes and allowing deployment of
+YuniKorn as the sole scheduler in a cluster with much greater confidence.
+
+Separate docker images are created for the scheduler. The traditional YuniKorn scheduler is built as
+`scheduler-{version}` while the new plugin version is built as `scheduler-plugin-{version}`. Either can be
+deployed interchangeably into a Kubernetes cluster with the same helm charts by customizing the scheduler
+image to deploy.
+
+## Entrypoints
+
+The existing shim `main()` method has been relocated to `pkg/cmd/shim/main.go`, and a new `main()` method
+under `pkg/cmd/schedulerplugin/main.go` has be created. This method instantiates the default Kubernetes
+scheduler and adds YuniKorn to it as a set of plugins. It also modifies the default scheduler CLI argument
+parsing to add YuniKorn-specific options. When the YuniKorn plugin is created, it will launch an instance
+of the existing shim / core schedulers in the background, sync all informers, and start the normal YuniKorn
+scheduling loop.
+
+## Shim Scheduler Changes
+
+In order to cooperate with the default scheduler, the shim needs to operate slightly differently when in
+plugin mode. These differences include:
+
+ - In `postTaskAllocated()`, we don’t actually bind the Pod or Volumes, as this is the responsibility of
+   the default scheduler framework. Instead, we track the Node that YK allocated for the Node in an
+   internal map, dispatch a new BindTaskEvent, and record a `QuotaApproved` event on the Pod.
+ - In `postTaskBound()`, we update the Pod’s state to `QuotaApproved` as this will cause the default scheduler
+   to re-evaluate the pod for scheduling (more on this below).
+ - In the scheduler cache, we track pending and in-progress pod allocations, and remove them if a pod is
+   removed from the cache.
+
+## Plugin Implementation
+
+To expose the entirety of YuniKorn functionality, we implement three of the Scheduling Framework Plugins:
+
+### PreFilter
+
+PreFilter plugins are passed a reference to a Pod and return either `Success` or `Unschedulable`, depending
+on whether that pod should be considered for scheduling.
+
+For the YuniKorn implementation, we first check the Pod to see if we have an associated `applicationID`
+defined. If not, we immediately return `Success`, which allows us to delegate to the default scheduler for
+non-batch workloads.
+
+If an `applicationId` is present, then we determine if there is a pending pod allocation (meaning the
+YuniKorn core has already decided to allocate the pod). If so, we return `Success`, otherwise `Unschedulable`.
+Additionally, if an in-progress allocation is detected (indicating that we have previously attempted to
+schedule this pod), we trigger a `TaskFailed` event for the YuniKorn core so that the pod will be sent back
+for scheduling later.
```

Comment:
I found this is quite confusing. If we have previously attempted to schedule a pod, and in PreFilter we see it again, we directly failed the allocation. Why do we consider this a failure? What if the actual binding took some time and is still ongoing? 

In another word, under what circumstance we can see a pod again in the PreFilter() plugin which was already scheduled by the default scheduler?

Reply:
If the default scheduler is unable to bind the pod, it will attempt to reschedule it, so we will see the same pod again in PreFilter(). In this case, we need to ensure that we release all the allocations in the core and send it back to the start, otherwise we risk being unable to schedule the pod at all. Note that this will only happen if the scheduler cannot bind a pod after it has already been cleared by Filter().

Reply:
Is this tested? If we mark the task as failed, I don't think the task will be rescheduled in YK.

Reply:
updated to reflect that we send 'RejectTask' instead

Reply:
After some long discussion. The current doc is aligning with the current implementation. But for this part, especially for retry logic, we need to do more tests and ensure the behavior is correct. If in the future we find anything incorrect that needs to be fixed, we will revisit the doc. Thx